### PR TITLE
Introduce Ansible console test

### PR DESCRIPTION
**Pull request by pdostal**
This test is single machine but uses **SSH** and **SUDO** as those are default.

- Related ticket: [poo#116935](https://progress.opensuse.org/issues/116935)
- Discovered bug: [bsc#1204182](https://bugzilla.opensuse.org/show_bug.cgi?id=1204182)
- Verification run: [Tumbleweed](https://pdostal-server.suse.cz/tests/2632), [Leap 15.4](https://pdostal-server.suse.cz/tests/2605), [Leap 15.3](https://pdostal-server.suse.cz/tests/2633), [SLE 15-SP4](https://pdostal-server.suse.cz/tests/2742), [SLE 15-SP3](https://pdostal-server.suse.cz/tests/2747)

### CODE REVIEW

File: data/console/ansible/roles/test/meta/main.yaml
Comment (grisu48) [inline data/console/ansible/roles/test/meta/main.yaml:10]:
```suggestion
    - name: opensuse
      versions:
        - 15.3
        - 15.4
        - tumbleweed
    - name: sles
      versions:
        - 12SP3
        - 12SP4
        - 12SP5
        - 15
        - 15SP1
        - 15SP2
        - 15SP3
        - 15SP4
```

For the suggested scheme see e.g. https://galaxy.ansible.com/api/v1/platforms/?search=sles

File: data/console/ansible/roles/test/tasks/main.yaml
Comment (grisu48) [inline data/console/ansible/roles/test/tasks/main.yaml:43]: `ed` or `htop`? The name of the task and the package name are inconsistent.

File: data/console/ansible/roles/test/templates/README.txt.j2
Comment (grisu48) [inline data/console/ansible/roles/test/templates/README.txt.j2:1]: AWESOME! :pineapple: (just a comment, mark as resolved when read :wink: )

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:20]: Leftover?

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:24]: Leftover?

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:31]: Leftover?

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:70]: Many leftovers?

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:80]: Is this necessary? Isn't the `zypper` module available by default?

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:109]: `assert_script_run` when executed in the `post_fail_hook` might have unintended side-effects (`die` while cleanup) - I should use `script_run` instead, which doesn't has this disadvantage.

**Review by grisu48 — COMMENTED** [review]
Good work Pavel! Some suggestions but overall nice job!

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:109]: If I remember correctly @okurz had opinion about this but I kinda forgot it. Oliver can you help? :smile:

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:80]: It's not necessary.

File: tests/console/ansible.pm
Comment (foursixnine) [inline tests/console/ansible.pm:26]: Please create the file in `sudoers.d` instead, as it would help with the cleanup step

File: products/alp/main.pm
Comment (foursixnine) [inline products/alp/main.pm:39]: schedule this at the end of `load_common_tests`

File: schedule/functional/extra_tests_textmode.yaml
Comment (foursixnine) [inline schedule/functional/extra_tests_textmode.yaml:103]: schedule right after `tar` tests, do note that Tumbleweed also schedules salt, and I'm not sure if that would bring unwanted consequences

Comment (pdostal) [reply schedule/functional/extra_tests_textmode.yaml:103]: I'm sure it will not and if yes then I will fix this test module to behave nicely :-)

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:26]: That is cool suggestion

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:91]: Actually why not :-) Let's also test that it is properly uninstallable :-)

File: lib/opensusebasetest.pm
Comment (b10n1k) [inline lib/opensusebasetest.pm:1334]: The parameter in the function is positional. Please revert

File: tests/console/ansible.pm
Comment (b10n1k) [inline tests/console/ansible.pm:60]: s/un/in/

File: tests/console/ansible.pm
Comment (b10n1k) [inline tests/console/ansible.pm:42]: Do you mean `systemctl 'enable --now sshd';` `start` will start the service but it wont same as enable.

**Review by b10n1k — COMMENTED** [review]
This seems a good example for wheels. or I would prefer to have the ansible playbooks in separate repo. in additional, i think this test make sense in multi-machine job, and i would expect some smoke test against `ansible all --list-hosts`. But i guess none of them is priority on this PR.

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:42]: No. It's just wrong comment :smile:

File: lib/opensusebasetest.pm
Comment (pdostal) [reply lib/opensusebasetest.pm:1334]: It's not. In the `select_serial_terminal()` there is also `$self` which makes `select_serial_terminal(0)` not work but after my change it works.

File: data/console/ansible/roles/test/meta/main.yaml
Comment (okurz) [reply data/console/ansible/roles/test/meta/main.yaml:10]: @grisu48 you really want to introduce a secluded definition file that needs to be updated with every SLE+Leap service pack?

**Comment by pdostal** [comment]
Wheels currently don't work.

What benefit do you see in multi machine?

File: lib/opensusebasetest.pm
Comment (b10n1k) [reply lib/opensusebasetest.pm:1334]: they are positional. The first expects the instance and then the value is assigned to the second variable.

Comment (b10n1k) [reply lib/opensusebasetest.pm:1334]: i still dont see any reason to not fix that.

**Comment by b10n1k** [comment]
> Wheels currently don't work.

Are we sure? i thought that it was fine. 
> 
> What benefit do you see in multi machine?

I am just saying that it makes sense in multi-machines. I doubt that there are significant benefits. Maybe catch some firewall, networking stuff in such setup.

File: lib/opensusebasetest.pm
Comment (pdostal) [reply lib/opensusebasetest.pm:1334]: When I revert it and call `select_user_serial_terminal()` it will log me in as **root** - not as **user**.
The `select_user_serial_terminal()` is not used anywhere else in this repository and is broken.

So can you tell me how should I fix this?

**Comment by pdostal** [comment]
> > Wheels currently don't work.
> 
> Are we sure? i thought that it was fine.

Yes, see: [poo#117622](https://progress.opensuse.org/issues/117622) and [poo#118903](https://progress.opensuse.org/issues/118903)
> 
> > What benefit do you see in multi machine?
> 
> I am just saying that it makes sense in multi-machines. I doubt that there are significant benefits. Maybe catch some firewall, networking stuff in such setup.

It may be nice to create another firewalld multimachine test. But as much as I love multimachine tests I know that they are expensive - harder to debug, harder to schedule, harder to run on various backends. I want this test module to be universal and easy to schedule.

But multimachine networking test would be nice :smile: Thanks for suggestions Yiannis!

File: lib/opensusebasetest.pm
Comment (pdostal) [reply lib/opensusebasetest.pm:1334]: Here is the verification run where I reverted just that one line: https://pdostal-server.suse.cz/tests/2197#step/ansible/30

Comment (b10n1k) [reply lib/opensusebasetest.pm:1334]:
```From 591b0c115d53e4d095c0e86542ef0a81de60f4f7 Mon Sep 17 00:00:00 2001
From: Ioannis Bonatakis <ybonatakis@suse.com>
Date: Mon, 17 Oct 2022 20:05:34 +0200
Subject: Fix select_user_serial_terminal

Signed-off-by: Ioannis Bonatakis <ybonatakis@suse.com>
---
 lib/opensusebasetest.pm | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/opensusebasetest.pm b/lib/opensusebasetest.pm
index 9478fffe2..937332a6b 100644
--- a/lib/opensusebasetest.pm
+++ b/lib/opensusebasetest.pm
@@ -1331,7 +1331,8 @@ The choice is made by BACKEND and other variables.
 =cut
 
 sub select_user_serial_terminal {
-    select_serial_terminal(root => 0);
+    my ($self) = @_;
+    $self->select_serial_terminal(0);
 }
 
 =head2 upload_coredumps
-- 
2.38.0```

Comment (pdostal) [reply lib/opensusebasetest.pm]: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15725

**Review by okurz — CHANGES_REQUESTED** [review]
The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a feature request entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

In particular for adding multiple files and a new test I suggest to be a bit more considerative regarding the test maintenance effort. I am not even clear on the need of this test and how deep it should go. I am in particular interested in any research about already existing test coverage of ansible so that it's clear what missing test coverage the openQA test can cover

File: data/console/ansible/roles/test/meta/main.yaml
Comment (grisu48) [reply data/console/ansible/roles/test/meta/main.yaml:10]: I would actually remove the `platforms` section for exactly this reason @okurz 
But if it is there, then we should do it properly.

Comment (grisu48) [reply data/console/ansible/roles/test/meta/main.yaml:10]: @pdostal your call - feel free to remove the `plaforms` section as a whole if you agree that we should not have this maintenance burden. The choice is up to you, I'm fine with your current approach as well :slightly_smiling_face:

File: tests/console/ansible.pm
Comment (grisu48) [inline tests/console/ansible.pm:47]: Leftover?

File: data/console/ansible/hosts
Comment (grisu48) [inline data/console/ansible/hosts:2]:
```suggestion
localhost ansible_connection=ssh
```

Following https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15684#issuecomment-1281118815 you could try to explicitly set `ansible_connection=ssh` so that ansible uses ssh. In this setup there should not be much of a difference between ansible running against localhost or another machine. It does not cover all use cases (e.g. dependencies on the remote host might still be a problem), but it would for sure cover the network dependent ansible connection.

Just a suggestion.

**Review by grisu48 — APPROVED** [review]
LGTM after editing the commit message according to https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15684#pullrequestreview-1145275474

**Comment by pdostal** [comment]
> LGTM after editing the commit message according to [#15684 (review)](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15684#pullrequestreview-1145275474)

I already did so. If you're not satisfied with my commit message, please make a suggestion.

**Review by okurz — APPROVED** [review]
nice!

File: tests/console/ansible.pm
Comment (foursixnine) [inline tests/console/ansible.pm:166]: I think stopping the Ansible service is needed too, isn't it? other than I'm pretty cool here :D

**Review by foursixnine — APPROVED** [review]

File: tests/console/ansible.pm
Comment (pdostal) [reply tests/console/ansible.pm:166]: Not valid, there is no systemd unit involved :smile:

**Pull request merged**
